### ISSUE TITLE
Grouped Window List shows 'ghost' applications

### ISSUE THREAD

**Issue by LinuxOnTheDesktop** [linuxmint/cinnamon#8955]
```
 * Cinnamon version (cinnamon --version) 4.2.4
   - Please specify if you are using the daily builds PPA (https://launchpad.net/~linuxmint-daily-build-team/+archive/ubuntu/daily-builds). I am not.
 * Distribution - (Mint 17.2, Arch, Fedora 25, etc...) Mint 19.2
 * Graphics hardware *and* driver used: Intel UHD Graphics 620 3.0 Mesa 19.0.8
 * 32 or 64 bit: 64
 * Attach /home/<username>/.xsession-errors, or /var/log/syslog: I attach both.
[xsesh.txt](https://github.com/linuxmint/cinnamon/files/3843484/xsesh.txt)
[syslog.txt](https://github.com/linuxmint/cinnamon/files/3843500/syslog.txt)

 ```

**Issue**

Sometimes Grouped Window List (GWL) presents an application as running after that application has closed - or at least seems to have closed. Sometimes the problem  takes this form: (1a) a pinned application (e.g. Thunderbird) continues to show the 'running' indicator (the horizontal bar) even when it is not running. At other times, (1b) an icon for a non-pinned application (e.g. the Update Manager) persists after closing. Sometimes (2) clicking the panel icon produces something with the appearance of a small titlebar. Whether or not 2 obtains, clicking anything - the icon, the titlebar-thing if present - does nothing.

**Steps to reproduce**

The problem is intermittent and I do not know the exact cause. It tends to occur with one application or another at least once in every session tha I use my computer. The problem seems to be new, or at least much more common that it used to be (some two months or so ago).

**Expected behaviour**

Clearly, all of this - icons showing for applications that are neither running nor pinned, pinned applications showing as running when they are not, applications showing as running not responding - just should not happen.

**Other information**

Here are two other outstanding and I believe gross GWL problems: (1) #8578; (2) #8796.

**Comment by jaszhix2** [comment]
Just a reminder, the previous maintainer of GWL (me - @jaszhix) has been banned since June, and have not heard from Clem or anyone part of the organization since then. I did not abandon GWL, it was work I was proud of, and petty, spiteful, politics ruined it for everyone else.

**Comment by LinuxOnTheDesktop** [comment]
It seems to me that Cinnamon had a severe need for something along the lines of GWL, i.e. for a proper taskbar / dock. That is why, despite this bug and the other two that I mention, I retain GWL. (Well, that and the fact that no third-party docks work well on Cinnamon - not even, in my experience anyway, Plank.) Those bugs affect me daily and I imagine that same holds for other people.

**Comment by jaszhix2** [comment]
Unfortunately they don't care about it, they care more about keeping their information proprietary behind a slack wall, they would rather ban a loose-lipped developer who might misrepresent them on accident, than have a maintainer work on software. You're getting the desktop you deserve by being complacent.

**Comment by LinuxOnTheDesktop** [comment]
@jaszhix2 

Complacent? I have become notorious for reporting problems. And I've tried a lot of other DEs and found them even more problematic than Cinnamon. But enough. Let's keep this bug report as a bug report. I see that it has been tagged as 'bug': that's a start.

**Comment by jaszhix2** [comment]
You're reporting the bugs, but nobody is going to fix this, it's been 5 months since I've been banned and nobody's touched this code. Get over yourself, this is just as contextually important as your mundane bug report.

**Comment by icarter09** [comment]
@jaszhix2  I know you are a good developer, but that is neither here nor there. If you have things you want to vent, hit me up on slack and I'll be more than happy to talk to you. But please stop with the unnecessary comments on this or any other bug ticket. I'm not trying to make you mad, but Github is not the place for this.

@LinuxOnTheDesktop thank you for taking the time to do the bug report. I have seen your name on many a bug/feature request and I value your input/opinion on both bug reports and feature requests. It is very much appreciated.

**Comment by jaszhix2** [comment]
@icarter09 you do realize I'm banned from all the slacks right? This is actually a appropriate place for me to talk about why these bugs still are present. This is not me venting, this is me rightfully pointing out why bugs are piling up for gwl. Besides applying labels to issues, what are you really contributing?

**Comment by icarter09** [comment]
@jaszhix2 here's my email icarter1391@gmail.com 

Create a separate slack channel or email me.

**Comment by jaszhix2** [comment]
I don't owe you any communications, I helped mentor you into becoming more involved with this organization and now I regret it.

**Comment by jaszhix2** [comment]
There's no creating a separate slack channel, that's not how slack works. I'm banned from the development slack site, the public development slack site, and the design slack site. I'm going to keep commenting on gwl issues because gwl for all intents and purposes is a product of my labor, and I don't like seeing it being neglected.

**Comment by icarter09** [comment]
@jaszhix icarter1391@gmail.com

You are right. You don't owe me anything. Just wanted to see if I could help out in anyway is all. I appreciate you mentoring me and liked working with you.

**Comment by jaszhix2** [comment]
After @brownsr offered to call me and never did, I'm not talking to any of you guys privately unless it's about unbanning me because that's where the conversation would start if there's any ethical integrity left here

**Comment by brownsr** [comment]
@jaszhix2 I did call a couple of times Jason, just got voicemail.  I think I left a message once, but it was a while ago, and I have the memory of a goldfish.

**Comment by jaszhix2** [comment]
I never received a phone call from anyone outside the US this year, so probably a wrong number.

**Comment by blueray453** [comment]
I am using Linux Mint 19.2, All updated. 

When several windows are open and I put mouse over the group icon (ex. Nemo/Firefox icon), it does not show the thumbnails. When I click on the group icon, it minimize and maximize only one window.

I then have to restart my pc to get the default behaviour (see the preview to open the appropriate window).

I am not actually sure what is causing this issue, and do not understand how to create bug report.

**Comment by icarter09** [comment]
@blueray453 

https://github.com/linuxmint/cinnamon/issues/new

**Comment by lestcape** [comment]
> Just a reminder, the previous maintainer of GWL (me - @jaszhix) has been banned since June, and have not heard from Clem or anyone part of the organization since then. I did not abandon GWL, it was work I was proud of, and petty, spiteful, politics ruined it for everyone else.

@jaszhix2 That happens when there are not difference between a third-party develop and the Mint one. It's to me a little strange that was specially you who are affecting for this now. When I did not agree with the new way to handled the extensions using a Mint git, you was a defender of that position. I don't know if karma exists, but well...

I think at less you understand now, that there is a fine line between what is right and what is not right enough. Cross it is really-really easy. Do not misunderstand me, I strongly agree with you that your should have a way to continues working in GWL from outside. The only difference is that I realized before that things like this could happen.

Anyway, the worst is that there are others who still do not see the problem. I hope, will not ending like you or me, investing a lot of time for nothing.

**Comment by leigh123linux** [comment]
That's enough!

**Comment by clefebvre** [comment]
There's a big different between you and him @lestcape. You're not banned from anything. You're able to voice disagreement without insulting everyone and you don't explode and libel everyone the minute you feel offended. You decided to stop contributing because of a decision we made.

The situation with Jason is completely different. He was asked to leave because he insulted his team mates. He was exposed on reddit because he lied about us on reddit. He was banned from github temporarily because he used github to spam us. He was filtered in my email because he spammed it as well. He was banned permanently because he escaped ban and showed even after months of no interaction between the team and him and he was still willing to harm himself if it could harm us. There's no way he'll ever be welcome to interact with us again. If he was smart he'd have moved on months ago. Don't compare yourself to him.

PS: I'm still persuaded the move towards git spices wasn't just right, it was needed. I'm happy to talk about it again though if I can convince you or at least explain why we did it.

To Jason, I don't know if you're still spamming my junk mail. Do stop to avoid this ban though. You were asked to leave before of this attitude you're having right now. You can decide not to learn from it, that's your choice, but do respect the fact that we no longer want to interact with you.

**Comment by clefebvre** [comment]
@LinuxOnTheDesktop we do care about GWL of course, it's our default window list. I've never run into this issue (the state continue to indicate the program is running even after the app is gone), but I would encourage you to find a sure way to reproduce it.

If it happens again, can you make sure the process of the app is dead? Say it's thunderbird, with a "ps -ef" for instance.

**Comment by lestcape** [comment]
> The situation with Jason is completely different.

I really don't want to know about personal problems or difficulty humans interactions. This is not my point. Regardless of who has the reason, the possibility of a conflict is there. My point then is that we need an easy way where both part can continues making code to the users, also if the conflict is irremediable. That way is making a distinction between what is of one and what is of another. But when both things are merged is really difficult find a good solution for both sides in a conflict.

Anyway, my greatest anger was not with the change itself, but with how it was done. But what was done, was done. I also understand that subsequently you tried to remedy things and I will said thank you for seeing the problem and correcting it.

> I'm still persuaded the move towards git spices wasn't just right, it was needed. I'm happy to talk about it again though if I can convince you or at least explain why we did it.

But i think is completely understandable the reason of move the **official** installation of Cinnamon extensions to a place that can be controlled. What is not so understandable is that in parallel there is no way to install extensions from a GUI source other than the official one at a user's decision (for example, a git repository of an external developer). 

I fully agree that you should ensure a good quality of code in the product you provide to the users, but you should not block the possibility that a user want or not to take the risk of install software from a non secure source of an external developer.

Probably, the installation of an extension directly from a git source, as a user decision is to much. It might even be enough to add an option in the cinnamon-settings to install an extension from a zip file or at least from the command line, but non of this options exist yet.

More software for Mint will always be good for Mint, right? So, why add restrictions to what users can do? It's not better warring hims about the possible risk of install non official software and done?

This is not the place to continues talking about this issue. Do you think it's ok open a new issue about install extensions from a non official source then?

**Comment by LinuxOnTheDesktop** [comment]
> @LinuxOnTheDesktop we do care about GWL of course, it's our default window list. I've never run into this issue (the state continue to indicate the program is running even after the app is gone), but I would encourage you to find a sure way to reproduce it. // If it happens again, can you make sure the process of the app is dead? Say it's thunderbird, with a "ps -ef" for instance.

I continue, of late, not to experience the ghosting problem. (Recently, acting on a tentative suggestion by Michael Webster - in #8796 -I took some newly committed GWL code and put it into my installation. Even before I did that, though, I had not seen the ghosting problem for at least about a month.) Still, the ghosting problem might reoccur (and someone else has the problem, it seems: see #8856, though I've only skimmed that page). Just now I have looked up `ps -ef` and I think I see how to use that command. But perhaps - because I don't have the problem currently - this issue should be closed (which would still leave #8856).

**Comment by collinss** [comment]
@lestcape There is currently a way to install 3rd party applets manually, (though not currently through a gui). There's a script that will install extensions via command-line. It's primarily for applet devs, so we don't really advertise it, but it's there.

But there is a problem with doing it directly in cinnamon-settings. The user often doesn't see that there's a distinction between what's provided by cinnamon, and what's 3rd party, and even if we put a notice, many people wont read it, and those who do may not really understand that they need to be careful, so I don't see how we can add this functionality to cinnamon itself without increasing risk to the user.

That doesn't mean that we can't find a way that works for everyone. One possibility would be to do it via ppa. When you add a ppa, you might not know where it's coming from still, but you at least know that it's from a different source than the os, and so you need to be a bit more careful. Also, someone could set up their own server and provide their own gui application that would allow the installation of 3rd party extensions. Neither of these would require much (if any) change on the cinnamon end, and could probably work out of the box, just using the internal mechanisms within cinnamon itself (such as the script I mentioned above).

**Comment by lestcape** [comment]
> There is currently a way to install 3rd party applets manually, (though not currently through a gui). There's a script that will install extensions via command-line. It's primarily for applet devs, so we don't really advertise it, but it's there.

Nice, well done!!

> But there is a problem with doing it directly in cinnamon-settings. The user often doesn't see that there's a distinction between what's provided by cinnamon, and what's 3rd party, and even if we put a notice, many people wont read it, and those who do may not really understand that they need to be careful, so I don't see how we can add this functionality to cinnamon itself without increasing risk to the user.

Yes, what you said is a description of the problem that should be discussed. I understand that is not easy like just add an option and done. Always all we do have repercussions and need to be analyses, but as you already mention:

> That doesn't mean that we can't find a way that works for everyone

> One possibility would be to do it via ppa. When you add a ppa, you might not know where it's coming from still, but you at least know that it's from a different source than the os, and so you need to be a bit more careful. Also, someone could set up their own server and provide their own gui application that would allow the installation of 3rd party extensions. Neither of these would require much (if any) change on the cinnamon end, and could probably work out of the box, just using the internal mechanisms within cinnamon itself (such as the script I mentioned above).

If there's a script that will install extensions via command-line, all will be possible then. But to be honest, this never was the problem. I can install an application that redirect all the system call to cinnamon-settings to my version of the cinnamon-settings and then handle the installation by my self.

The question to me is more about if a solution like that will be officially accepted or not? You know, i don't want to play to the mouse and the cat, where i try to install external extensions and you try that my code will no work. So far what I understand was that non of you want that user should install extensions from outside, because security reason. You are tell me that this is not in that way?

Also, you really prefer that some one else make the way to install external extensions? It's not much more secured to the user if the installer is also provide by Cinnamon itself? I don't understand the direction here then. Is confused to me and i really don't want to do what none of cinnamon dev want to have, because that finally will  be ending in nothing.

**Comment by clefebvre** [comment]
We don't need to prevent the user from installing 3rd party software, all we need to do is be responsible for the software we distribute (and that includes all spices coming from the spice website). In other words, if a spice is downloaded by the user from somewhere else, I've no problem with making it easy to install. In that situation the user downloaded the software himself, so it's pretty obvious to him it is 3rd party and the responsibility lies with him.

Most DEs (outside of KDE, Cinnamon) don't even have a store, or a repo, or whatever we call it. All they have is a + button, people click to point to a .zip, or a directory. We could add that easily and let people install applets/desklets/extensions locally.

**Comment by lestcape** [comment]
@clefebvre to me this is a properly solution. If someone wants something better they can do it by himself because cinnamon-settings allows modules. I think then, this will send the right message that users to, the command line is not an option for user, but this is one in my opinion.

**Comment by LinuxOnTheDesktop** [comment]
Closed, because I reported the problem and I have it no longer. One can still add to this thread, though.

**Comment by mattvalley** [comment]
@LinuxOnTheDesktop Could this issue be the same behavior that I am seeing in #8779 ?  In that issue, I am seeing the "ghost" application icons in the panel when I close an application (using the panel context menu or thumbnail close button) while the application is minimized.  I have been able to reliably reproduce the ghost application icons in that issue.

**Comment by clefebvre** [comment]
@lestcape Cool, it shouldn't be hard to add this in 4.6. I added the idea to the roadmap.

**Comment by lestcape** [comment]
No, it should not be hard. It's more less the same of https://github.com/linuxmint/cinnamon/pull/7420 and adapt it to what we need. 

If you will add this at roadmap, just take in concideration that the new feature is not useful only for install non official extensions, is also useful for install official extensions in a computer that is offline.

**Comment by blueray453** [comment]
@lestcape I think what @clefebvre meant by
> it shouldn't be hard to add this in 4.6

Is that there are a lot more high priority stuff to be dealt with until 4.6 

However, I do not know about your internal politics, and also did not understand what feature you are talking about. But if it is really an important feature to you and @clefebvre really wants it included, you can lend him an helping hand.

It is really an exciting time for Linux Mint and we all should focus on the big picture. It is healthy to have a debate about the prospect or future about the project, but when it is all said and done, we should follow the leader. This is how we stay together, this is how we survive.

**Comment by lestcape** [comment]
@blueray453 i never haven a leader and I've always stayed on the periphery to not have one. Also, I have never used Mint and i'm currently not use Cinnamon even. I think It's also important first know what is the conversation to have then an strong opinion like your own.  Can you help us here then? We are only trying to create a better place to you as a Linux user. Have a good day.

**Comment by blueray453** [comment]
@lestcape sorry, I made a fool of myself. I should not have jumped into the conversation without reading previous comments carefully.

**Comment by jaszhix3** [comment]
> There's a big different between you and him @lestcape. You're not banned from anything. You're able to voice disagreement without insulting everyone and you don't explode and libel everyone the minute you feel offended. You decided to stop contributing because of a decision we made.

This is ironically libel. 

> The situation with Jason is completely different. He was asked to leave because he insulted his team mates. 

Example?

> He was exposed on reddit because he lied about us on reddit. 

Actually, I exposed the ban on Reddit. You then decided to make up a story about me "abusing" team mates with no example. That is **libel**.

> He was banned from github temporarily because he used github to spam us. 

What happened first though? I believe I was banned well before I attempted to reason with you guys. I was banned right before you wanted to revert a lot of my work.

> He was filtered in my email because he spammed it as well. 

Spam would imply the messages were excessive. I don't think sending one email once every two months is excessive. This just shows you had no intention of reasoning with me.

> He was banned permanently because he escaped ban and showed even after months of no interaction with the team that he was still willing to harm himself if it could harm us. 

This makes no sense. Criticizing Linux Mint doesn't harm me in any way. In fact, the work I've done here has no contributed to my professional development in any way. Developing UIs for Linux desktops is niche, and doesn't translate well into what I do for a living.

> There's no way he'll ever be welcome to interact with us again. If he was smart he'd have moved on months ago. Don't compare yourself to him.

This condescension is exactly why I moved on, but I'm allowed to be angry. You completely disrespected me and my work, and continue to do so. 

> To Jason, I don't know if you're still spamming my junk mail. Do stop to avoid this ban though. 

You just said the ban is permanent, so there's no reason for me to not create a third account to responded your libel.

> You were asked to leave because of this attitude you're having right now. You can decide not to learn from it, that's your choice, but do respect the fact that we no longer want to interact with you.

The only thing I've learned from this experience is to not trust you. Any dissent is considered a bad attitude. You have no tolerance for public disagreement, and that's the whole point of the barriers you put up between the core team and the public (invite-only slack the public is left out on). If you don't want to interact with me, you can stop attacking my personality to justify your over-reactive moderation.
